### PR TITLE
*: set store labels using pd api.

### LINKED ISSUE #667: Feature request: Allow PD to add/modify topological location label to TiKV dynamically
When deploying TiDB cluster in kubernetes, we can't add topological location label to TiKV before it's deployed because we don't know where it'll be deployed.
So we should allow PD to add/modify topological location labels to TiKV dynamically through PD api, for example with pd-ctl
```
store show label 1  # get store 1's label
store set label 1 zone=<zone>,rack=<rack>,host=<host> # set store 1's label with zone,rack,host
```
For binary deployment which is deterministic, label TiKV through command line arguments or configuration file is still needed.

### PR DESCRIPTION
**Pull request by disksing**
Fix #667.

Using pd-ctl:
```
> store label 1 key1 value1
```
Using api:
```
curl -X POST -d '{"key1":"value1", "key2":"value2"}' http://127.0.0.1:2379/pd/api/v1/store/1/label
```

### CODE REVIEW

File: server/api/router.go [hunk 1]
```diff
@@ -58,6 +58,7 @@ func createRouter(prefix string, svr *server.Server) *mux.Router {
 	storeHandler := newStoreHandler(svr, rd)
 	router.HandleFunc("/api/v1/store/{id}", storeHandler.Get).Methods("GET")
 	router.HandleFunc("/api/v1/store/{id}", storeHandler.Delete).Methods("DELETE")
+	router.HandleFunc("/api/v1/store/{id}/label", storeHandler.SetLabels).Methods("POST")
```
Comment (siddontang) [inline server/api/router.go]: PUT is not allowed here?

Comment (disksing) [inline server/api/router.go]: No. Keep it the same with other APIs.

File: server/replication.go [hunk 2]
```diff
@@ -63,29 +63,14 @@ func (r *Replication) GetDistinctScore(stores []*storeInfo, other *storeInfo) fl
...
-				return 0
-			}
-			id2 := other.getLocationID(keys)
-			if len(id2) == 0 {
-				return 0
-			}
-			if id1 != id2 {
-				score += levelScore
-			}
+	for _, s := range stores {
```
Comment (siddontang) [inline server/replication.go]: should use another PR

File: server/store.go [hunk 3]
```diff
@@ -131,16 +131,31 @@ func (s *storeInfo) getLabelValue(key string) string {
...
 
-func (s *storeInfo) getLocationID(keys []string) string {
-	id := ""
-	for _, k := range keys {
-		v := s.getLabelValue(k)
-		if len(v) == 0 {
-			return ""
+// compareLocation compares 2 stores' labels and returns at which level their
+// locations are different. It returns -1 if they are at the same location.
+func (s *storeInfo) compareLocation(other *storeInfo, labels []string) int {
```
Comment (siddontang) [inline server/store.go]: any test?

File: server/replication.go (hunk 2 above)
Comment (disksing) [inline server/replication.go]: ok.

File: server/store.go (hunk 3 above)
Comment (disksing) [inline server/store.go]: It's a utility function used by `GetDistinctScore`, which is reasonably tested.

**Comment by disksing** [comment]
PTAL @siddontang @nolouch

File: pdctl/command/store_command.go [hunk 4]
```diff
@@ -47,6 +49,16 @@ func NewDeleteStoreCommand() *cobra.Command {
 	return d
 }
 
+// NewLabelStoreCommand returns a label subcommand of storeCmd.
+func NewLabelStoreCommand() *cobra.Command {
+	l := &cobra.Command{
+		Use:   "label <store_id> <key> <value>",
+		Short: "set a store's label value",
```
Comment (huachaohuang) [inline pdctl/command/store_command.go]: Seems this command actually add labels to the store.
Can we add sub commands to get/set/add labels to a store?

File: server/api/router.go (hunk 1 above)
Comment (huachaohuang) [inline server/api/router.go]: Can we use PUT to replace the original labels and POST to add new labels?

File: server/api/store_test.go [hunk 5]
```diff
@@ -118,6 +121,28 @@ func (s *testStoreSuite) TestStoreGet(c *C) {
 	checkStoresInfo(c, []*storeInfo{info}, s.stores[:1])
 }
 
+func (s *testStoreSuite) TestStoreLabel(c *C) {
```
Comment (huachaohuang) [inline server/api/store_test.go]: Should test "merge" case with some original labels.

File: server/api/store.go [hunk 6]
```diff
@@ -154,6 +154,42 @@ func (h *storeHandler) Delete(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+
+	var input map[string]string
+	if err := readJSON(r.Body, &input); err != nil {
+		h.rd.JSON(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+	var labels []*metapb.StoreLabel
+	for k, v := range input {
```
Comment (nolouch) [inline server/api/store.go]: may the order of the labels be inconsistent with the origin? though the `pd-ctl` command just set one label. better to add a comment.

File: server/api/router.go (hunk 1 above)
Comment (disksing) [inline server/api/router.go]: I don't think we have to distinguish replace and new.

File: server/api/store.go (hunk 6 above)
Comment (disksing) [inline server/api/store.go]: The order of the labels won't matter. Labels are compared by the order they are listed in pd's config file.

File: pdctl/command/store_command.go (hunk 4 above)
Comment (disksing) [inline pdctl/command/store_command.go]: I suppose no need to mind the difference of set and add. And the `store id` api will show labels, so seems unnecessary to add a `get` neither.

**Comment by huachaohuang** [comment]
LGTM

**Pull request merged**